### PR TITLE
Fix `torch.compile` on `nn.module` instead of on `LightningModule`

### PR DESCRIPTION
**Pull request by tesfaldet**
## What does this PR do?

Related to this issue https://github.com/Lightning-AI/lightning/issues/17177.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tesfaldet** [comment]
Tests can be fixed by merging this PR https://github.com/ashleve/lightning-hydra-template/pull/585

File: requirements.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # --------- pytorch --------- #
+python>=3.10,<3.11
```
Comment (ashleve) [inline requirements.txt:2]: Package setup fails on this line

Comment (tesfaldet) [reply requirements.txt:2]: Can you give more information? Or rather, can you tell me how you tested for this?

Comment (ashleve) [reply requirements.txt:2]: Sorry for confusion. Check out github actions tests
image

Comment (tesfaldet) [reply requirements.txt:2]: Apologies for the hold-up, I'll get to this soon! Been a busy several weeks.

**Review by ashleve — APPROVED** [review]

File: environment.yaml [hunk 2]
```diff
@@ -21,6 +21,7 @@ channels:
 # compatibility is usually guaranteed
 
 dependencies:
+  - python=3.10
```
[the remark this answers is not in the document]
Comment (tesfaldet) [reply environment.yaml:24]: Yeah, unfortunately

[the remark this answers is not in the document]
Comment (ofarajun) [reply environment.yaml:24]: Are you sure?  I see [here](https://pytorch.org/blog/pytorch-2.0-release/#deprecation-of-cuda-116-and-python-37-support-for-pytorch-20) that only python 3.8 is required.

**Comment by ResearchDaniel** [comment]
Revert this fix as it has been fixed in PyTorch?
https://github.com/Lightning-AI/lightning/issues/17177#issuecomment-1666023472
It is more convenient with a global switch

**Comment by tesfaldet** [comment]
> Revert this fix as it has been fixed in PyTorch? [Lightning-AI/lightning#17177 (comment)](https://github.com/Lightning-AI/lightning/issues/17177#issuecomment-1666023472) It is more convenient with a global switch

I noticed this recently as well. I'm currently testing it for myself. I'm not sure which version of PyTorch fixes this problem as I don't see a version mentioned in the comment https://github.com/Lightning-AI/lightning/issues/17177#issuecomment-1666023472 but I'll try to find it.

File: environment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tesfaldet) [reply environment.yaml:24]: I forget exactly the error I got, so I'll try to recreate it to show that 3.10 is required, for posterity's sake.

**Pull request merged**